### ISSUE TITLE
Manually implemented setters are unsafe (potential over-release & potential mutation of a set value)

### ISSUE THREAD

**Issue by d-ronnqvist** [AzureAD/azure-activedirectory-library-for-objc#557]
For example, consider this setter

```
- (void)setUserInformation:(ADUserInformation *)userInformation
{
    SAFE_ARC_RELEASE(_userInformation);
    _userInformation = userInformation;
    SAFE_ARC_RETAIN(_userInformation);
    [self calculateHash];
}
```

If the same object were to be set twice (meaning that `_userInformation` already points to `userInformation`). Then that object would be released before it was retained, potentially causing it to deallocate, which would mean that the `retain` message would be sent to the deallocated object.

See http://www.cocoawithlove.com/2010/06/assign-retain-copy-pitfalls-in-obj-c.html

**Comment by d-ronnqvist** [comment]
Further, some manually implemented setters for immutable versions of classes that have mutable counterparts (e.g. NSString, NSArray, NSDictionary, ...) retain the new value, even thought the property is declared as copying. For example:

```
- (void)setResource:(NSString *)resource
{
    SAFE_ARC_RELEASE(_resource);
    _resource = resource;
    SAFE_ARC_RETAIN(_resource);
    [self calculateHash];
} 
```

which is defined as:

```
@property (copy) NSString* resource;
```

This means that even thought the property is correctly defined as copying, the settler is still susceptible to the type of problems that copy accessors solve. 

Again, see: http://www.cocoawithlove.com/2010/06/assign-retain-copy-pitfalls-in-obj-c.html

**Comment by RPangrle** [comment]
A number of these were fixed in #558 but I'll go over the code to see if I can find more.
